packages/cli/src/commands/onboard/discovery/providers/github/GithubRepository.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { graphql } from '@octokit/graphql';
+import {
+  Repository as GraphqlRepository,
+  Tree as GraphqlTree,
+  Blob as GraphqlBlob,
+} from '@octokit/graphql-schema';
+import { Repository, RepositoryFile } from '../types';
+import { RepositoryResponse } from './graphql';
+import { GithubFile } from './GithubFile';
+
+export class GithubRepository implements Repository {
+  readonly #client: typeof graphql;
+  readonly #repo: RepositoryResponse;
+  readonly #org: string;
+  #files: Promise<RepositoryFile[]> | undefined;
+
+  constructor(client: typeof graphql, repo: RepositoryResponse, org: string) {
+    this.#client = client;
+    this.#repo = repo;
+    this.#org = org;
+  }
+
+  get url(): string {
+    return this.#repo.url;
+  }
+
+  get name(): string {
+    return this.#repo.name;
+  }
+
+  get owner(): string {
+    return this.#org;
+  }
+
+  get description(): string | undefined {
+    return this.#repo.description;
+  }
+
+  files(): Promise<RepositoryFile[]> {
+    this.#files ??= this.#doGetFiles();
+    return this.#files;
+  }
+
+  async #doGetFiles(): Promise<RepositoryFile[]> {
+    const query = `query RepoFiles($owner: String!, $name: String!) {
+        repository(owner: $owner, name: $name) {
+          object(expression: "HEAD:") {
+            ... on Tree {
+              entries {
+                name
+                type
+                object {
+                  ... on Blob {
+                    byteSize
+                    text
+                    isBinary
```

Comment:
![image](https://user-images.githubusercontent.com/110673802/235874427-36cb6b13-e23f-41fb-a375-1cad10e89768.png)

Are we fetching every file in a repository, _including its contents_?
I'm worried about performance here, especially since the only usage of this data is to get the contents of `package.json`.

Perhaps we should replace this function with something like:
```
async fetchPackageJson: Promise<text | undefined> {
    // fetch package.json, returning "undefined" if it doesn't exist
    // throw an error if a request fails for another reason, etc
    return unparsedJson
}
```

Reply:
Yeah this  make sense to refactor a bit. Your earlier suggestion of getting a single file instead should sounds like a good approach. I'll give that a try!

Reply:
https://github.com/backstage/backstage/pull/17624/commits/94dcf05da54fc197c1a155dc8bd12a9dad17b50a